nucypher/characters/lawful.py

```diff
@@ -1716,7 +1806,41 @@ def interface_info_with_metadata(self):
     # Re-Encryption
     #
 
-    def _reencrypt(self, kfrag: VerifiedKeyFrag, work_order: 'WorkOrder', alice_verifying_key: PublicKey):
+    def verify_kfrag_authorization(self,
+                                   alice: Alice,
+                                   kfrag: KeyFrag,
+                                   signed_writ: bytes,
+                                   work_order: 'WorkOrder'
+                                   ) -> VerifiedKeyFrag:
+        from nucypher.policy.orders import WorkOrder  # TODO: resolve ciruclar dependency
+
+        writ_hrac, writ_kfrag_checksum, writ_signature = WorkOrder.signed_writ_splitter(signed_writ)
+        reconstructed_writ = writ_hrac + writ_kfrag_checksum
+
+        try:
+            self.verify_from(alice, reconstructed_writ, signature=writ_signature)
+        except InvalidSignature:
+            raise Policy.Unauthorized  # This isn't from Alice (relayer).
+
+        if writ_hrac != work_order.hrac:  # Funky Workorder
+            raise Policy.Unauthorized  # Bob, what the *hell* are you doing?
+
+        kfrag_checksum = keccak_digest(bytes(kfrag))[:WRIT_CHECKSUM_SIZE]
+        if kfrag_checksum != writ_kfrag_checksum:
+            raise Policy.Unauthorized  # Bob, Seriously?
+
+        try:
+            verified_kfrag = kfrag.verify(verifying_pk=alice.stamp.as_umbral_pubkey())
+        except VerificationError:
+            raise Policy.Unauthorized  # WTF, Alice did not generate these KFrags.
+
+        if writ_hrac in self.revoked_policies:
+            # Note: This is only an off-chain and in-memory check.
+            raise Policy.Unauthorized  # Denied
```

Comment:
Do we want to differentiate the reason for these raised `Policy.Unauthorized` exceptions (L1830, L1835, L1839) by contextualizing with a string message? Or is it unsafe to leak the reason?

Reply:
Feels like it - the one raised from the `VerificationError` (L1835) may be interesting...

Reply:
Alternatively, we could do sub-classes of Policy.Unauthorized if differentiation is worth it.

Reply:
Filed https://github.com/nucypher/nucypher/issues/2740.